### PR TITLE
add cdn datasource

### PR DESCRIPTION
**Pull request by crab21**
$ make testacc TESTARGS=" -run=TestAccTencentCloudCdnDomains"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCdnDomains -timeout 120m
?   	github.com/terraform-providers/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-tencentcloud/gendoc	1.438s [no tests to run]
=== RUN   TestAccTencentCloudCdnDomains
--- PASS: TestAccTencentCloudCdnDomains (212.71s)
PASS
ok  	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud	214.189s
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]

### CODE REVIEW

File: tencentcloud/data_source_tc_cdn_domains.go [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+				ValidateFunc: validateAllowedStringValue(CDN_SERVICE_TYPE),
+				Description:  "Service type of Acceleration domain name. The available value include `web`, `download` and `media`.",
+			},
+			"full_url_cache": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "Whether to enable full-path cache.",
+			},
+			"origin_pull_protocol": {
+				Type:         schema.TypeBool,
```
Comment (Himer) [inline tencentcloud/data_source_tc_cdn_domains.go:49]: Type:         schema.Typestring?

File: tencentcloud/data_source_tc_cdn_domains.go [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+			value = "on"
+		} else {
+			value = "off"
+		}
+
+		domainFilterMap["full_url_cache"] = value
+	}
+
+	if len(domainFilterMap) == 0 {
+		return nil
```
Comment (Himer) [inline tencentcloud/data_source_tc_cdn_domains.go:230]: 没有filter 需要拉出所有的list

File: tencentcloud/extension_cdn.go [hunk 3]
```diff
@@ -31,6 +31,8 @@ const (
 
 	CDN_HOST_NOT_FOUND      = "ResourceNotFound.CdnHostNotExists"
 	CDN_DOMAIN_CONFIG_ERROE = "FailedOperation.CdnConfigError"
+
+	CDN_DATASOURCE_NAME_DOMAIN = "domain"
```
Comment (Himer) [inline tencentcloud/extension_cdn.go:35]: CDN_DATASOURCE_NAME_DOMAIN不能定义个常量

File: tencentcloud/common.go [hunk 4]
```diff
@@ -240,3 +240,31 @@ func IsContains(array interface{}, value interface{}) bool {
 		return reflect.DeepEqual(array, value)
 	}
 }
+
+func UnderlineToHump(underline string) (humpValue string) {
```
Comment (Himer) [inline tencentcloud/common.go:244]: 如果只有cdn使用  需要加个CDN前缀  并放在service_cdn这个文件里边

Comment (Himer) [reply tencentcloud/common.go:244]: 而且UnderlineToHump有更简单的形式   遍历underline byte来判断&append操作就可以了

File: tencentcloud/data_source_tc_cdn_domains.go [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+			"domain": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Name of the acceleration domain.",
+			},
+			"service_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateAllowedStringValue(CDN_SERVICE_TYPE),
+				Description:  "Service type of Acceleration domain name. The available value include `web`, `download` and `media`.",
```
Comment (Himer) [inline tencentcloud/data_source_tc_cdn_domains.go:41]: Domain name service type
web: static acceleration
download: download acceleration
media: streaming VOD acceleration

File: tencentcloud/data_source_tc_cdn_domains.go [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+)
+
+func dataSourceTencentCloudCdnDomains() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceTencentCloudCdnDomainsRead,
+		Schema: map[string]*schema.Schema{
+			"domain": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Name of the acceleration domain.",
```
Comment (Himer) [inline tencentcloud/data_source_tc_cdn_domains.go:35]: 这些描述 从cdn的国际站粘贴复制即可  能不自己翻译的就不自己翻译  https://intl.cloud.tencent.com/document/api/228/31739#BriefDomain

File: tencentcloud/data_source_tc_cdn_domains.go [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+	_ = d.Set("status", domainConfig.Status)
+
+	if *domainConfig.CacheKey.FullUrlCache == CDN_SWITCH_OFF {
+		_ = d.Set("full_url_cache", false)
+	} else {
+		_ = d.Set("full_url_cache", true)
+	}
+
+	origins := make([]map[string]interface{}, 0, 1)
+	origin := make(map[string]interface{}, 8)
```
Comment (Himer) [inline tencentcloud/data_source_tc_cdn_domains.go:265]: map 无需如此精确大小

File: tencentcloud/data_source_tc_cdn_domains.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+	origins = append(origins, origin)
+	_ = d.Set("origin", origins)
+
+	httpsConfig := make(map[string]interface{}, 7)
+	httpsConfig["https_switch"] = domainConfig.Https.Switch
+	httpsConfig["http2_switch"] = domainConfig.Https.Http2
+	httpsConfig["ocsp_stapling_switch"] = domainConfig.Https.OcspStapling
+	httpsConfig["spdy_switch"] = domainConfig.Https.Spdy
+	httpsConfig["verify_client"] = domainConfig.Https.VerifyClient
+	_ = d.Set("https_config", httpsConfig)
```
Comment (Himer) [inline tencentcloud/data_source_tc_cdn_domains.go:283]: set全部搞错了   cdn_domain_list没有设置 https_config在  cdn_domain_list里边无法这样设置

File: tencentcloud/data_source_tc_cdn_domains.go [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+
+	if len(domainFilterMap) == 0 {
+		return nil
+	}
+
+	var domainConfig *cdn.DetailDomain
+	var errRet error
+	err := resource.Retry(readRetryTimeout, func() *resource.RetryError {
+		domainConfig, errRet = cdnService.DescribeDomainsConfigByFilters(ctx, domainFilterMap)
+		if errRet != nil {
```
Comment (Himer) [inline tencentcloud/data_source_tc_cdn_domains.go:237]: DescribeDomainsConfigByFilters只返回一个域名? 这是不对的..... 要返回所有满足条件的cdn列表

File: tencentcloud/data_source_tc_cdn_domains.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+	httpsConfig["ocsp_stapling_switch"] = domainConfig.Https.OcspStapling
+	httpsConfig["spdy_switch"] = domainConfig.Https.Spdy
+	httpsConfig["verify_client"] = domainConfig.Https.VerifyClient
+	_ = d.Set("https_config", httpsConfig)
+
+	tags, errRet := tagService.DescribeResourceTags(ctx, CDN_SERVICE_NAME, CDN_RESOURCE_NAME_DOMAIN, region, domain.(string))
+	if errRet != nil {
+		return errRet
+	}
+	_ = d.Set("tags", tags)
```
Comment (oliverpei) [inline tencentcloud/data_source_tc_cdn_domains.go:289]: 这里的set都没set进去

File: tencentcloud/provider.go [hunk 11]
```diff
@@ -536,6 +536,7 @@ func Provider() terraform.ResourceProvider {
 			"tencentcloud_cos_bucket_object":            dataSourceTencentCloudCosBucketObject(),
 			"tencentcloud_cos_buckets":                  dataSourceTencentCloudCosBuckets(),
 			"tencentcloud_cfs_file_systems":             dataSourceTencentCloudCfsFileSystems(),
+			"tencentcloud_cdn_domains":                  dataSourceTencentCloudCdnDomains(),
```
Comment (oliverpei) [inline tencentcloud/provider.go:539]: 没加目录

File: tencentcloud/service_tencentcloud_cdn.go [hunk 12]
```diff
@@ -88,3 +88,38 @@ func (me *CdnService) StartDomain(ctx context.Context, domain string) error {
...
+			logId, request.GetAction(), request.ToJsonString(), err.Error())
+		errRet = err
+		return
+	}
+
+	if len(response.Response.Domains) > 0 {
+		domainConfig = response.Response.Domains[0]
+	}
+	return
+}
```
Comment (oliverpei) [inline tencentcloud/service_tencentcloud_cdn.go:125]: 这个查的是个列表，不能只取第一个，参考其他的data source 怎么写的

File: tencentcloud/data_source_tc_cdn_domains.go [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	cdn "github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/cdn/v20180606"
+)
+
+func dataSourceTencentCloudCdnDomains() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceTencentCloudCdnDomainsRead,
+		Schema: map[string]*schema.Schema{
+			"domain": {
```
Comment (Himer) [inline tencentcloud/data_source_tc_cdn_domains.go:32]: 只有domain就可以了吧 其他的其实不需要filter

File: website/docs/d/cdn_domain.html.markdown [hunk 14]
```diff
@@ -0,0 +1,64 @@
+---
```
Comment (likexian) [inline website/docs/d/cdn_domain.html.markdown:1]: 这个文档是多余的，可以删除

**Comment by crab21** [comment]
add cdn domain result fileds.
***

$ make testacc TESTARGS=" -run=TestAccTencentCloudCdnDomains"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCdnDomains -timeout 120m
?   	github.com/terraform-providers/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-tencentcloud/gendoc	(cached) [no tests to run]
=== RUN   TestAccTencentCloudCdnDomains
--- PASS: TestAccTencentCloudCdnDomains (134.77s)
PASS
ok  	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud	135.981s
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]

File: tencentcloud/service_tencentcloud_cdn.go [hunk 15]
```diff
@@ -88,3 +89,55 @@ func (me *CdnService) StartDomain(ctx context.Context, domain string) error {
...
+		if sdkErr, ok := err.(*errors.TencentCloudSDKError); ok {
+			if sdkErr.Code == CDN_HOST_NOT_FOUND {
+				return
+			}
+		}
+		log.Printf("[CRITAL]%s api[%s] fail, request body [%s], reason[%s]\n",
+			logId, request.GetAction(), request.ToJsonString(), err.Error())
+		errRet = err
+		return
+	}
```
Comment (oliverpei) [inline tencentcloud/service_tencentcloud_cdn.go:120]: 没有处理分页，只能查询100条数据

File: tencentcloud/service_tencentcloud_cdn.go [hunk 16]
```diff
@@ -88,3 +89,55 @@ func (me *CdnService) StartDomain(ctx context.Context, domain string) error {
...
+				i++
+			}
+			continue
+		}
+
+		humpValue += string(underline[i])
+	}
+
+	return
+}
```
Comment (oliverpei) [inline tencentcloud/service_tencentcloud_cdn.go:143]: 这个函数可以不需要，调用DescribeDomainsConfigByFilters 的参数 filterMap 的key直接就填写成你需要的名称就可以

File: tencentcloud/data_source_tc_cdn_domains.go [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+				Optional:    true,
+				Description: "Used to save results.",
+			},
+			"domain_list": {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "An information list of cdn domain. Each element contains the following attributes:",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"resource_id": {
```
[the remark this answers is not in the document]
Comment (oliverpei) [reply tencentcloud/data_source_tc_cdn_domains.go:71]: 改成id吧

**Comment by crab21** [comment]
remove redundancy code and add `offset` field.
***
$ make testacc TESTARGS=" -run=TestAccTencentCloudCdnDomains"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCdnDomains -timeout 120m
?   	github.com/terraform-providers/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-tencentcloud/gendoc	0.921s [no tests to run]
=== RUN   TestAccTencentCloudCdnDomains
--- PASS: TestAccTencentCloudCdnDomains (121.35s)
PASS
ok  	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud	121.978s
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]

**Pull request merged**